core/src/main/java/cucumber/runtime/Runtime.java

```diff
@@ -1,66 +1,251 @@
+import cucumber.runtime.model.FeatureLoader;
 import gherkin.events.PickleEvent;
 
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 
 /**
  * This is the main entry point for running Cucumber features from the CLI.
  */
 public class Runtime {
 
-    private final Runner runner;
+    private final ExitStatus exitStatus;
+
+    private final RuntimeOptions runtimeOptions;
+
+    private final RunnerSupplier runnerSupplier;
     private final Filters filters;
     private final EventBus bus;
-    private final FeatureCompiler compiler = new FeatureCompiler();
     private final FeatureSupplier featureSupplier;
     private final Plugins plugins;
+    private final ExecutorService executor;
 
-    public Runtime(Plugins plugins,
-                   EventBus bus,
-                   Filters filters,
-                   RunnerSupplier runnerSupplier,
-                   FeatureSupplier featureSupplier
-    ) {
+    public Runtime(final Plugins plugins,
+                   final RuntimeOptions runtimeOptions,
+                   final EventBus bus,
+                   final Filters filters,
+                   final RunnerSupplier runnerSupplier,
+                   final FeatureSupplier featureSupplier,
+                   final ExecutorService executor) {
 
         this.plugins = plugins;
+        this.runtimeOptions = runtimeOptions;
         this.filters = filters;
         this.bus = bus;
-        this.runner = runnerSupplier.get();
+        this.runnerSupplier = runnerSupplier;
         this.featureSupplier = featureSupplier;
+        this.executor = executor;
+        this.exitStatus = new ExitStatus(runtimeOptions);
+        exitStatus.setEventPublisher(bus);
     }
 
     public void run() {
-        List<CucumberFeature> features = featureSupplier.get();
+        final List<CucumberFeature> features = featureSupplier.get();
         bus.send(new TestRunStarted(bus.getTime()));
         for (CucumberFeature feature : features) {
             feature.sendTestSourceRead(bus);
         }
 
-        StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
+        final StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
+        runnerSupplier.get().reportStepDefinitions(stepDefinitionReporter);
 
-        runner.reportStepDefinitions(stepDefinitionReporter);
-
-        for (CucumberFeature cucumberFeature : features) {
-            runFeature(cucumberFeature);
+        final FeatureCompiler compiler = new FeatureCompiler();
+        for (CucumberFeature feature : features) {
+            for (final PickleEvent pickleEvent : compiler.compileFeature(feature)) {
+                if (filters.matchesFilters(pickleEvent)) {
+                    executor.execute(new Runnable() {
```

Comment:
If I see this correctly, that will simply swallow any exceptions thrown while executing a pickle - i.e it will log them, dispose the executor thread and then execute the next pickle in a new executor thread.
I have seen this behaviour because there's a bug in my ConcurrentEventListener. It doesn't make the cucumber run fail!

Reply:
Can you make an issue for that? Ideally we catch all exceptions in the runner and fail the pickle.

Reply:
See https://github.com/cucumber/cucumber-jvm/issues/1628